scipy-1.0/ckdtree.tex

```diff
@@ -2,22 +2,20 @@
 periodic boundary conditions was added. A periodic boundary condition is typically 
 used in computer simulations of physical processes.
 
-% TODO: please check / improve literature citations used here
-% what about the time complexity of the kNN implementation?
-In 2015, we enhanced \texttt{cKDTree.query} with a $k$ nearest neighbors search
-parameter. This is an efficient operation\cite{Sproull:1991:RNS:3118219.3118331} 
-because it generates and returns a data structure that may be only a fraction 
-of the size of the full neighbor list. A C++ implementation is provided that releases
-the Python global interpreter lock for the neighbor search process, uses
-a memory pool to allocate and automatically reclaim \texttt{structs}, and
-heapsort to sort the $k$ nearest neighbors distances, even in the case
-of periodic boundary conditions. When $k$ is a non-contiguous list of nearest
-neighbors to account for (i.e., \texttt{[1, 3, 4]}), the intervening distances
-between the nearest neighbor ($k = 1$) and the farthest neighbor requested
-are all interrogated, but only those neighbors that are requested are retained
-in memory. Thus, the maximum heap size for this algorithm is conceptually determined
-as \texttt{np.arange(max(k))}.
+Specifically, in 2015, we enhanced \texttt{cKDTree.query} by reimplementing in C++,
+removing some of the algorithm memory leaks, and allowing 
+release of the GIL so that multiple threads may be used\cite{gh-4374}. In 2013,
+the time complexity of the k nearest neighbor search from \texttt{cKDTree.query}
+was approximately loglinear \cite{knn-jake}, and we've confirmed that this time
+complexity has been preserved in SciPy \texttt{1.0.0} (Figure~\ref{fig:knn-complexity}),
+but with generally improved efficiency (Figure~\ref{fig:asvbench}). The
+discussion in the originally-cited nearest neighbor search
+algorithm publication\cite{kdtree-search-algo}, suggests that there are
+at least some scenarios where $O(log n)$ search time complexity may
+be achieved, but the influence of dimensionality appears to be consistent
+with our empirical observations of approximately $0(n log n)$ performance.
```

Comment:
I think this is referring to the part in the reference:
"The seminal work in this area was by Friedman, Bentley, and Finkel [21] who showed that O(n) space and O(log n) query time are achievable for fixed dimensional spaces in the expected case for data distributions of bounded density through the use of kd-trees. There have been numerous variations on this theme. However, all known methods suffer from the fact that as dimension increases, either running time or space increase exponentially with dimension."
But before reading this in the reference, this part of the paper sounded to me like there might be a problem with our implementation. 
Is the intention of citing the paper to give evidence that O(n log n) is reasonable?

Reply:
I removed these specifics based on your proposed refactoring above, since "roughly loglinear" seems to flow nicely enough while preserving some concision.